### PR TITLE
[FIX] web_editor: ensure checklist changes are saved

### PR DESCRIPTION
**Pull request by dhba-odoo**
**Current behavior before PR:**

Clicking the checkbox triggered the `mousedown` event, where the `preventDefault`
function prevented the list from gaining focus. Consequently, without any focused 
element, clicking the breadcrumb (`o_back_button`) to navigate back failed to trigger
the `blur` event, resulting in unsaved changes.

**Desired behavior after PR is merged:**

After removing `preventDefault` from the `mousedown` event, clicking the checkbox will
now set focus on it. As a result, when the breadcrumb `o_back_button` is clicked, it 
will trigger the `blur` event, leading to the execution of `onWysiwygBlur` in `html_field`,
which will commit(save) our changes.

task-3346155

### CODE REVIEW

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 1]
```diff
@@ -629,8 +628,6 @@ export class HtmlField extends Component {
         const rating = shouldToggleOff ? 0 : previousStars.length + 1;
 
         let starsId = $(node).parent().attr('id');
-        starsId = starsId && starsId.replace('checkId-', '');
-        starsId = parseInt(starsId || '0');
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/backend/html_field.js:633]: @Zinston Could you have a look at those changes please ? I'm very confused by the removal of those lines as it seems to partially undo what was done at https://github.com/odoo/odoo/pull/83137/commits/1f61f0d397eee6be2cbcf4042a1c15fa07063fb2#diff-c1e8fb6ac2e4699027311d5c71fb2679f25ca75fbb8cbac00f79d77b55d7a13dR526.

Comment (Zinston) [reply addons/web_editor/static/src/js/backend/html_field.js:633]: This is confusing the hell out of me.

With these changes:

Say we have `<star id="checkId-1"/>`.
1. `let starsId = $(node).parent().attr('id');` so `starsId === 'checkId-1'` (and `let` should be `const`).
2. We do an RPC to `/web_editor/stars` with `starsId = 'checkId-1'`
3. In `update_stars`: `stars_widget = htmlelem.find(".//span[@id='checkId-%s']" % starsId)`
4. That's equivalent to `stars_widget = htmlelem.find(".//span[@id='checkId-checkId-1']")`
5. Therefore `stars_widget is None`
6. We return just below in `if stars_widget is None: return value`.

So we never check the star. I don't see how this could ever work.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 2]
```diff
@@ -2657,3 +2657,27 @@ export function cleanZWS(node) {
...
+
+/**
+ * check click or mouse down event occurred inside the checkbox box.
+ *
+ * @param {Node} node
+ * @returns {boolean}
+ */
+export function isMouseInsideCheckbox(ev) {
+    const node = ev.target;
+    const beforStyle = window.getComputedStyle(node, ':before');
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2669]: spell check `beforeStyle`

**Comment by shsa-odoo** [comment]
@dhba-odoo  This branch has Conflicting files

File: addons/web_editor/controllers/main.py [hunk 3]
```diff
@@ -265,7 +265,8 @@ def update_stars(self, res_model, res_id, filename, starsId, rating):
                 star.set('class', classname)
             star_index += 1
 
-        value = etree.tostring(htmlelem[0][0], encoding='utf-8', method='html')[5:-6]
+        value_bytes = etree.tostring(htmlelem[0][0], encoding='utf-8', method='html')[5:-6]
+        value = value_bytes.decode('utf-8')
```
Comment (Zynton) [inline addons/web_editor/controllers/main.py:269]: Wouldn't the `[5:-6]` need to be done on the string rather that on the bytestring then?

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 4]
```diff
@@ -604,7 +605,6 @@ export class HtmlField extends Component {
         let checklistId = $(ev.target).attr('id');
         checklistId = checklistId && checklistId.replace('checkId-', '');
         checklistId = parseInt(checklistId || '0');
-
```
Comment (Zynton) [inline addons/web_editor/static/src/js/backend/html_field.js:607]: No need to remove this.

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 5]
```diff
@@ -155,6 +148,14 @@ export class HtmlField extends Component {
                 }
             })();
         });
+        onMounted(() => {
+            const $editable = this.wysiwyg && this.wysiwyg.getEditable();
+            if ($editable && $editable.length) {
+                $editable.off('.checklistBinding');
+                $editable.on('mousedown.checklistBinding', 'ul.o_checklist > li', this._onReadonlyClickChecklist.bind(this));
```
Comment (Zynton) [inline addons/web_editor/static/src/js/backend/html_field.js:155]: Why did you change from click to mousedown?

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 6]
```diff
@@ -138,14 +139,6 @@ export class HtmlField extends Component {
                         this._qwebPlugin.sanitizeElement(this.readonlyElementRef.el);
                         // Ensure all external links are opened in a new tab.
                         retargetLinks(this.readonlyElementRef.el);
-
-                        const hasReadonlyModifiers = Boolean(this.props.record.isReadonly(this.props.fieldName));
```
Comment (Zynton) [inline addons/web_editor/static/src/js/backend/html_field.js:142]: Just to make sure I understand. This used to restrict the feature to when the field was readonly. Now it's also active when it's in edit but the field isn't dirty so that it will save whether the field eventually gets focused or not (which is the bug you're fixing here). Correct?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js (hunk 2 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2778]: I agree. Even if it's a copy-paste, you should take the opportunity to correct the typo.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 7]
```diff
@@ -4335,6 +4322,7 @@ export class OdooEditor extends EventTarget {
                     star.classList.toggle('fa-star-o', true);
                     star.classList.toggle('fa-star', false);
                 };
+                ev.preventDefault();
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4325]: Why?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 8]
```diff
@@ -2766,3 +2766,27 @@ export function cleanZWS(node) {
...
+    }
+    style1.right = style1.left + parseInt(beforStyle.getPropertyValue('width'), 10);
+    style1.bottom = style1.top + parseInt(beforStyle.getPropertyValue('height'), 10);
+
+    const isMouseInsideCheckboxBox =
+        ev.offsetX >= style1.left &&
+        ev.offsetX <= style1.right &&
+        ev.offsetY >= style1.top &&
+        ev.offsetY <= style1.bottom;
+    return isMouseInsideCheckboxBox;
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2791]:
```suggestion
    return offsetX >= style1.left &&
        ev.offsetX <= style1.right &&
        ev.offsetY >= style1.top &&
        ev.offsetY <= style1.bottom;
```

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 9]
```diff
@@ -2766,3 +2766,27 @@ export function cleanZWS(node) {
         .filter(node => node.nodeType === Node.TEXT_NODE)
         .forEach(node => node.nodeValue = node.nodeValue.replace(/\u200B/g, ''));
 }
+
+/**
+ * check click or mouse down event occurred inside the checkbox box.
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2771]:
```suggestion
 * Return true if click or mousedown event occurred inside the checkbox box.
 * False otherwise.
```

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 10]
```diff
@@ -595,7 +596,7 @@ export class HtmlField extends Component {
         }
     }
     async _onReadonlyClickChecklist(ev) {
-        if (ev.offsetX > 0) {
+        if (!isMouseInsideCheckbox || this._isDirty()) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/backend/html_field.js:599]: You're just checking that the function `isMouseInsideCheckbox` was properly imported?

**Review by Zynton — CHANGES_REQUESTED** [review]

File: addons/web_editor/controllers/main.py (hunk 3 above)
Comment (dhba-odoo) [reply addons/web_editor/controllers/main.py:269]: I've checked both, and they give the same result. So, we can go with either one.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 7 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4325]: Without the `preventDefault`, it saves changes made in the star element even without calling `_onReadonlyClickStar`. Initially, I followed the same approach for the checklist, and it worked; removing `preventDefault` for the checklist saved it even if `_onReadonlyClickChecklist` was not called.

However, after reading DMO's log note, I discovered that there's already a function that can save the checklist when first clicked. So, I revised my fix.
and i have removed `preventDefault` here to avoid potential future issues because of this.

File: addons/web_editor/static/src/js/backend/html_field.js (hunk 5 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js:155]: I called the `mousedown` event to ensure that it is triggered before the editing value is updated.

File: addons/web_editor/static/src/js/backend/html_field.js (hunk 6 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js:142]: yes!

File: addons/web_editor/static/src/js/backend/html_field.js (hunk 10 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: I have updated it!

File: addons/web_editor/static/src/js/backend/html_field.js (hunk 5 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js:155]: I guess that's an ok tradeoff.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 11]
```diff
@@ -4291,21 +4292,7 @@ export class OdooEditor extends EventTarget {
...
-            style1.bottom = style1.top + parseInt(beforStyle.getPropertyValue('height'), 10);
-
-            const isMouseInsideCheckboxBox =
-                ev.offsetX >= style1.left &&
-                ev.offsetX <= style1.right &&
-                ev.offsetY >= style1.top &&
-                ev.offsetY <= style1.bottom;
-
-            if (isMouseInsideCheckboxBox) {
+            if (isMouseInsideCheckbox(ev)) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4295]:
```suggestion
        if (node.tagName == 'LI' && getListMode(node.parentElement) == 'CL' && isMouseInsideCheckbox(ev)) {
```
then outdent the contents.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 12]
```diff
@@ -4291,21 +4292,7 @@ export class OdooEditor extends EventTarget {
         const node = ev.target;
         // handle checkbox lists
         if (node.tagName == 'LI' && getListMode(node.parentElement) == 'CL') {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4294]:
```suggestion
        if (node.tagName == 'LI' && getListMode(node.parentElement) == 'CL' && isMouseInsideCheckbox(ev)) {
```
then remove the nested `if` and outdent its contents.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 13]
```diff
@@ -2766,3 +2766,26 @@ export function cleanZWS(node) {
         .filter(node => node.nodeType === Node.TEXT_NODE)
         .forEach(node => node.nodeValue = node.nodeValue.replace(/\u200B/g, ''));
 }
+
+/**
+ * Return true if click or mousedown event occurred inside the checkbox box.
+ * False otherwise.
+ *
+ * @param {Node} node
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2774]: The parameter is not called `node` and is not of type `Node`.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 14]
```diff
@@ -2766,3 +2766,26 @@ export function cleanZWS(node) {
...
+ * @returns {boolean}
+ */
+export function isMouseInsideCheckbox(ev) {
+    const node = ev.target;
+    const beforeStyle = window.getComputedStyle(node, ':before');
+    const style1 = {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2780]: Actually also while you're at it, maybe you could rename this to `style`. I don't get why it was called `style1` if there's no `style2`.

**Review by Zynton — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 12 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: done!

**Comment by Zynton** [comment]
@dhba-odoo Please let me know when the conflicts are resolved.

**Comment by dhba-odoo** [comment]
@Zynton Conflict has been resolved. Please have a look.

File: addons/web_editor/controllers/main.py [hunk 15]
```diff
@@ -265,7 +265,7 @@ def update_stars(self, res_model, res_id, filename, starsId, rating):
                 star.set('class', classname)
             star_index += 1
 
-        value = etree.tostring(htmlelem[0][0], encoding='utf-8', method='html')[5:-6]
+        value = etree.tostring(htmlelem[0][0], encoding='unicode', method='html')[5:-6]
```
Comment (Zynton) [inline addons/web_editor/controllers/main.py:268]: Commit message: remove the dot at the end of the title.

File: addons/web_editor/static/src/js/backend/html_field.js (hunk 10 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: It worries me that it passed testing (automated and human) while the first condition was always false. Could we add a test for this? I think another run of Testing IN/BE would be good after the changes.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 16]
```diff
@@ -2766,3 +2766,26 @@ export function cleanZWS(node) {
         .filter(node => node.nodeType === Node.TEXT_NODE)
         .forEach(node => node.nodeValue = node.nodeValue.replace(/\u200B/g, ''));
 }
+
+/**
+ * Return true if click or mousedown event occurred inside the checkbox box.
+ * False otherwise.
+ *
+ * @param {Object} ev
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:2774]: To be more precise:

```suggestion
 * @param {Event} ev
```

File: addons/web_editor/static/src/js/backend/html_field.js (hunk 10 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: I have considered creating a QUnit test case for this, but I am unable to execute the final step necessary for this task. This step involves navigating back from the current record and reopening it to verify if the checklist has been saved correctly.

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: I attempted to create a tour for this but encountered an issue. The problem occurs because the `tooltip` added during the tour process prevents the code responsible for saving the checklist from executing.
I've added a condition where if `_isDirty()` returns true, the code returns from the `_onReadonlyClickChecklist` function before the code responsible for saving the checklist is executed. and because of the `tooltip`, `_isDirty()` returns true.

Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: Maybe you can push the commit with the failing tour so I can take a look, because like this it's hard for me to tell what's to be done.

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: I have pushed the tour commit. You can take a look at it

Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: Your commit is supposed to be about checking a checkbox _in readonly mode_. Why is your tour testing _in edit mode_? If you were in readonly mode you shouldn't need worry about `isDirty`. Which begs the question: why did you add these `isDirty` conditions in the first place?

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: I had a discussion with @dmo-odoo in the task _lognote_ regarding _readonly_ and go to know that **from 16.0**, the `html_field` is only in readonly mode if it lacks edit rights. There's no scenario where it possesses edit rights but it can goes into readonly mode also. In our case, since we can edit in notes, I don't believe it can switch to readonly mode.

That is why i have use `_isDirty()` here. with this ,  I am determining that if it returns false, the page is in readonly mode because the user hasn't made any edits in the record. However, if it returns true, it indicates that the user has edited the record, so it can be considered as being in edit mode.

> Your commit is supposed to be about checking a checkbox in readonly mode. Why is your tour testing in edit mode? 

I don't think i am checking it in edit mode or i am missing something, because i have created a step to save the record before the checkbox is checked so to make sure that when checkbox is checked at that time it is not in edit mode.
or because of trigger it goes back to in edit mode?

Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: I think there's a misunderstanding here about _readonly_ and `_isDirty`. If `_isDirty` is true, that would indeed probably indicate that you're in edit mode because you're not supposed to be able to make changes in readonly, by definition. But the converse is not true! `_isDirty` being false **does not mean the record is _readonly_**: it just means that no change has been made since the last time the record was saved.

So you are in fact testing in edit mode.

Anyway, now having read again your discussion with @dmo-odoo in the task, and the steps to reproduce, I understand it's not really a question of readonly/edit but probably rather that clicking the checkbox doesn't focus the editor so that clicking on the breadcrumb after that doesn't cause a blur event, and nothing is saved. That or the check mark doesn't cause the record to register as dirty, but that would be strange. Can you check if the first hypothesis is true ? I'll take a look at your approach based on your answer. Thanks.

(Either way, to reproduce the steps, you need to leave the page before clicking on the checkbox.)

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: Yes. In this situation, the _blur_ event isn't being triggered, and it's because the `mousedown` event on the checkbox fails to focus the editor. This issue is likely caused by [this](https://github.com/odoo/odoo/blob/82ac6aee2702eb3d4ba3b01047e9cd3a3d66bfa5/addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js#L4348[this]). because after removing it, focus is properly set. As a result, both the _blur_ event is triggered when the back button is clicked, and the checklist is also successfully saved.

Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: Any idea why that line had been added? It seems like the right fix is indeed to remove it.

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: I don't know why it was added, but it has been there since this [PR](https://github.com/odoo-dev/odoo-editor/commit/721055e3afca2d5966069657634ce6a1381aed9b)

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: The tour is still not functioning properly. Upon investigation, it appears that dispatching `mousedown` and `click` events within the tour are not triggering the `focus` and `blur` events as expected.

Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: That might be, you may have to trigger them yourself.

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: Yes, it might work, but manually triggering those events isn't a good idea. If I were to do so, the tour would still proceed even if I were to remove my fix. This is because what I've done is removed the `ev.preventDefault` inside `mousedown`, which is responsible for preventing focus on the checklist when the checkbox is clicked.

Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: @dmo-odoo Does this ring a bell? @dhba-odoo is writing a tour that relies on dispatching `mousedown` and `click` events and they don't seem to impact the focus (or at least cause `focus` and `blur` events). Any chance you'd know of a workaround?

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: Maybe because the events are not trusted, like the arrow keys? What about calling focus manually just for the purpose of the test? Before doing that, check for other tests that might do the same, as I'd be surprised that this would be the first time.

Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: @dmo-odoo

> What about calling focus manually just for the purpose of the test?

That would kind of defeat the purpose of the test because the fix he's testing involves not preventing a focus event. But agreed on the other part (checking other tests) @dhba-odoo

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: @dmo-odoo @Zynton I've looked into the other test cases, and most of the time the mouseup event is triggered after the mousedown event, or manual focus and blur events are used. I have tried both approaches and only manually triggering focus and blur events works. However, this approach wouldn't work for our specific need because triggering focus and blur manually would still pass the test case without my fix too.

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: @dhba-odoo What happens if you try to reproduce the same case in a jsfiddle and manually trigger a mousedown or click in js? Do you get focus/blur events then ? If yes, then something in our code is preventing those events from firing and we need to find what. If no, then we can't test this bug due to that limitation. cc @Zynton

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/backend/html_field.js]: I checked on `JSFiddle`, and the same thing happens there too. When manually triggering a `mousedown` event, it correctly calls the focus event. However, when using dispatchEvent, only the `mousedown` event is triggered, and the focus event doesn't occur. The same problem occurs with the `click` and `blur` events as well.

Comment (Zynton) [reply addons/web_editor/static/src/js/backend/html_field.js]: I don't, I was going to say the same thing as you.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 17]
```diff
@@ -4417,7 +4417,6 @@ export class OdooEditor extends EventTarget {
 
             if (isMouseInsideCheckboxBox) {
                 toggleClass(node, 'o_checked');
-                ev.preventDefault();
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4420]: Could you update the commit message to explain why the `preventDefault` was creating the issue in the first place? What other piece of code somewhere else was prevented to run because of it?

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4334]: @dmo-odoo I have updated the commit message!

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:4334]: Thanks! 👍

**Comment by dmo-odoo** [comment]
@robodoo r+
